plugins/modules/net_tools/nmcli.py

```diff
@@ -746,53 +798,32 @@ def bool_to_string(boolean):
-            settings = dbus.Interface(proxy, "org.freedesktop.NetworkManager.Settings")
-        except dbus.exceptions.DBusException as e:
-            self.module.fail_json(msg="Unable to read Network Manager settings from DBus system bus: %s" % to_native(e),
-                                  details="Please check if NetworkManager is installed and"
-                                          "service network-manager is started.")
-        connection_paths = settings.ListConnections()
-        connection_list = []
-        # List each connection's name, UUID, and type
-        for path in connection_paths:
-            con_proxy = bus.get_object(service_name, path)
-            settings_connection = dbus.Interface(con_proxy, "org.freedesktop.NetworkManager.Settings.Connection")
-            config = settings_connection.GetSettings()
-
-            # Now get secrets too; we grab the secrets for each type of connection
-            # (since there isn't a "get all secrets" call because most of the time
-            # you only need 'wifi' secrets or '802.1x' secrets, not everything) and
-            # merge that into the configuration data - To use at a later stage
-            self.merge_secrets(settings_connection, config, '802-11-wireless')
-            self.merge_secrets(settings_connection, config, '802-11-wireless-security')
-            self.merge_secrets(settings_connection, config, '802-1x')
-            self.merge_secrets(settings_connection, config, 'gsm')
-            self.merge_secrets(settings_connection, config, 'cdma')
-            self.merge_secrets(settings_connection, config, 'ppp')
-
-            # Get the details of the 'connection' setting
-            s_con = config['connection']
-            connection_list.append(s_con['id'])
-            connection_list.append(s_con['uuid'])
-            connection_list.append(s_con['type'])
-            connection_list.append(self.connection_to_string(config))
-        return connection_list
+    @staticmethod
+    def list_to_string(lst):
+        return ",".join(lst or [""])
 
-    def connection_exists(self):
-        # we are going to use name and type in this instance to find if that connection exists and is of type x
-        connections = self.list_connection_info()
+    @staticmethod
+    def settings_type(setting):
+        if setting in ('bridge.stp',
+                       'bridge-port.hairpin-mode',
+                       'connection.autoconnect'):
+            return bool
+        elif setting in ('ipv4.dns',
+                         'ipv4.dns-search',
+                         'ipv6.dns',
+                         'ipv6.dns-search'):
+            return list
+        return None
 
-        for con_item in connections:
-            if self.conn_name == con_item:
-                return True
+    def connection_exists(self):
+        cmd = [self.nmcli_bin, 'con', 'show', self.conn_name]
```

Comment:
It is recommend to use command `nmcli con show` instead of `nmcli con show <conn_name>` for detecting if connection exists. Here're some reasons:
1. The performance of `nmcli con show <conn_name>` may slower than the former one, because all details will be dumped to the standard output.
2. Using `nmcli con show` could avoid most potential exceptions in parsing and displaying active connection data. 
3. Not only connection name, but also any other aliases (uuid, path, path), can be used to denote a connection in `nmcli con show <conn_id>`. But it's better only to match the connection name for determining the existence.

The result of connection list fetched from `nmcli con show` can also be cached in Nmcli object to avoid duplicated queries. If so, the cached data should be cleaned up in method `remove_connection()` and `create_connection()` as well.

Reply:
Exactly the type of feedback I was looking for!  I'll look into fixing the `connection_exists` implementation tonight.

Reply:
Fixed in ffeae0b8, but no caching.